### PR TITLE
Fix the confusing-zero case with NestedScrollView.

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
cc @collinjackson for review

**Comment by Hixie** [comment]
Fixes https://github.com/flutter/flutter/issues/10658

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 1]
```diff
@@ -863,3 +1141,503 @@ class _NestedOuterBallisticScrollActivity extends BallisticScrollActivity {
...
+  RenderSliverOverlapAbsorber createRenderObject(BuildContext context) {
+    return new RenderSliverOverlapAbsorber(
+      handle: handle,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, RenderSliverOverlapAbsorber renderObject) {
+    renderObject
+      ..handle = handle;
```
Comment (collinjackson) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Cascade operator looks kind of weird here if you're only setting one thing.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 2]
```diff
@@ -863,3 +1141,503 @@ class _NestedOuterBallisticScrollActivity extends BallisticScrollActivity {
...
+  RenderSliverOverlapInjector createRenderObject(BuildContext context) {
+    return new RenderSliverOverlapInjector(
+      handle: handle,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, RenderSliverOverlapInjector renderObject) {
+    renderObject
+      ..handle = handle;
```
Comment (collinjackson) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Cascade operator looks kind of weird here if you're only setting one thing.

**Review by collinjackson — APPROVED** [review]
This is a pretty big change, and it's hard for me to reason about the ways it might break, but it seems to behave intuitively in my testing. It exposes a lot of new API surface area for us to maintain, but I don't see a good way around that.

I wonder if it would make sense to have the `NestedScrollView` constructor adds the injectors/absorber automagically (maybe also other things like PageStorageKey and SafeArea), and then have a `NestedScrollView.custom` constructor that gives lower-level control over these advanced widgets if you don't like the default behavior. If you agree that this is a good idea, it could be done in a separate pull request.

File: examples/flutter_gallery/lib/demo/material/tabs_demo.dart [hunk 3]
```diff
@@ -141,13 +143,16 @@ class TabsDemo extends StatelessWidget {
...
-                  tabs: _allPages.keys.map((_Page page) => new Tab(text: page.label)).toList(),
+              new SliverOverlapAbsorber(
+                handle: NestedScrollView.sliverOverlapAbsorberHandleFor(context),
+                child: new SliverAppBar(
+                  title: const Text('Tabs and scrolling'),
+                  pinned: true,
+                  expandedHeight: 150.0,
+                  forceElevated: innerBoxIsScrolled,
+                  bottom: new TabBar(
+                    tabs: _allPages.keys.map((_Page page) => new Tab(text: page.label)).toList(),
```
[the remark this answers is not in the document]
Comment (Hixie) [reply examples/flutter_gallery/lib/demo/material/tabs_demo.dart]: Split most of the lines down. A few still exceed 80 chars but there's no clean way to split those.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 4]
```diff
@@ -142,20 +305,84 @@ class _NestedScrollViewState extends State<NestedScrollView> {
...
+      child: new Builder(
+        builder: (BuildContext context) {
+          return new _NestedScrollViewCustomScrollView(
+            scrollDirection: widget.scrollDirection,
+            reverse: widget.reverse,
+            physics: widget.physics != null
+                ? widget.physics.applyTo(const ClampingScrollPhysics())
+                : const ClampingScrollPhysics(),
+            controller: _coordinator._outerController,
+            slivers: widget._buildSlivers(context, _coordinator._innerController, _coordinator.hasScrolledBody),
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: done

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Yeah. This is done for consistence with all the other updateRenderObject methods in the framework.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 5]
```diff
@@ -863,3 +1141,503 @@ class _NestedOuterBallisticScrollActivity extends BallisticScrollActivity {
...
+
+/// A sliver that wraps another, forcing its layout extent to be treated as
+/// overlap.
+///
+/// The difference between the overlap requested by the [child] sliver and the
+/// overlap reported by this widget, called the _absorbed overlap_, is reported
+/// to the [SliverOverlapAbsorberHandle], which is typically passed to a
+/// [RenderSliverOverlapInjector].
+class RenderSliverOverlapAbsorber extends RenderSliver with RenderObjectWithChildMixin<RenderSliver> {
+  /// Create a slvire that absorbs overlap and reports it to a
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: fixed

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: as above

File: packages/flutter/test/rendering/mock_canvas.dart [hunk 6]
```diff
@@ -637,6 +654,11 @@ class _TestRecordingCanvasPatternMatcher extends _TestRecordingCanvasMatcher imp
     _predicates.add(new _FunctionPaintPredicate(#drawParagraph, <dynamic>[paragraph, offset]));
   }
 
+  @override
+  void shadow({ Iterable<Offset> includes, Iterable<Offset> excludes, Color color, double elevation, bool transparentOccluder }) {
+    _predicates.add(new _ShadowPredicate(includes: includes, excludes: excludes, color: color, elevation: elevation, transparentOccluder: transparentOccluder));
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/test/rendering/mock_canvas.dart]: i left this one... it's part of a bunch of more or less identical methods that just duplicate their arguments into constructor calls. It's probably easier to read if it's just skipped over as obviously mechanical code.

**Comment by Hixie** [comment]
Thanks for the review! Will land on green unless I hear otherwise.

**Comment by collinjackson** [comment]
lgtm

**Pull request merged**